pkg/iam/iam.go

```diff
@@ -0,0 +1,399 @@
+						}
+					}
+				}
+			}
+		}
+	}
+	return *userPolicyDocs
+}
+
+// GetPermissionsFromPolicyDoc parses a PolicyDocument and extracts
+// a list of effective Actions granted. It only counts Actions that
+// are granted to all resources (listed as "Resource": "*" in the
+// Policy Document).
+func GetPermissionsFromPolicyDoc(userPolicyDocs *[]string) []string {
+	// permissions is the refined collection of all permissions
+	// granted to this user through IAM.
+	var permissions []string
+
+	// Look through all Policy Documents associated with this user
+	// and gather a list of all permissions granted to the user.
+	logrus.Debugf("Found the following permissions applied to the current user:")
+	for _, p := range *userPolicyDocs {
+		u, err := url.QueryUnescape(p)
+		logrus.Debugf(u)
+		if err != nil {
+			logrus.Fatalf("Error decoding policy document: %s", p)
+		}
+
+		userPolicy := []byte(u)
+
+		// a stores the arbitrary json response received from AWS.
+		// It contains the IAM policy documents related to this user.
+		var a map[string]interface{}
+		err = json.Unmarshal(userPolicy, &a)
+		for _, statement := range a["Statement"].([]interface{}) {
+			elem := statement.(map[string]interface{})
+			if elem["Effect"] == "Allow" {
+				var singleAction bool
+				var allResources bool
+				_, ok := elem["Resource"].(string)
+				if ok {
+					if elem["Resource"] == "*" {
+						allResources = true
+					}
+				}
+				_, ok = elem["Action"].(string)
+				if ok {
+					singleAction = true
+				}
+				if allResources && singleAction {
+					singleact, ok := elem["Action"].(string)
+					if ok {
+						permissions = append(permissions, singleact)
+					}
+				} else if allResources {
+					for _, action := range elem["Action"].([]interface{}) {
+						act, ok := action.(string)
+						if ok {
+							permissions = append(permissions, act)
+						} else {
+							logrus.Fatalf("Error appending action %q to permissions list: %v", action, permissions)
+						}
+					}
+				}
+			}
+		}
+	}
+	return permissions
+}
+
+// CheckOffPermissionsByPrefix accepts a map of permissions
+// to be used as a checklist. It then marks all permissions 'true'
+// if they match the specified prefix.
+func CheckOffPermissionsByPrefix(checklist map[string]bool, prefix string) map[string]bool {
+	if prefix == "*" {
+		for k := range checklist {
+			checklist[k] = true
+		}
+	}
+	for k := range checklist {
+		match, _ := regexp.MatchString(prefix, k)
+		if match {
+			logrus.Debugf("Prefix %q matches checklist item %q", prefix, k)
+			checklist[k] = true
+		}
+	}
+	return checklist
+}
+
+// RunPermissionsCheck ties together all the above helper functions
+// to run a complete suite of AWS IAM permissions checks. This will
+// inspect permissions attached directly to users and groups, as well
+// as managed policies containing permissions granted to the user.
+func RunPermissionsCheck() {
+
+	// Create an IAM service client.
+	sess, err := session.NewSession(&aws.Config{})
```

Comment:
I am not quite familiar enough to know from what I'm seeing, but will this code work when your aws creds are set in the standard env vars?

Reply:
Yes, this should work with env var credentials. I took it right from the AWS docs, so it's the standard way to create a session, as far as I know. Though you bring up a good point. I'll test with  AWS env vars and leave out the `~/.aws/config` file to be sure. I was using both at the same time, since I believe both are needed to run the installer.

Reply:
To get creds from the usual places, I think you should follow [this pattern][1].

[1]: https://github.com/openshift/installer/blob/963d1a2dc8c659315957b0db0d8e1b178659f780/pkg/rhcos/ami.go#L24-L29

Reply:
Thanks, done.